lightning/src/routing/router.rs

```diff
@@ -1643,96 +1643,57 @@ where L::Target: Logger {
 		return Err(LightningError{err: "Failed to find a sufficient route to the given destination".to_owned(), action: ErrorAction::IgnoreError});
 	}
 
-	// Sort by total fees and take the best paths.
-	payment_paths.sort_unstable_by_key(|path| path.get_total_fee_paid_msat());
-	if payment_paths.len() > 50 {
-		payment_paths.truncate(50);
-	}
+	// Step (6).
+	let mut selected_route = payment_paths;
 
-	// Draw multiple sufficient routes by randomly combining the selected paths.
-	let mut drawn_routes = Vec::new();
-	let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 12]);
-	let mut random_index_bytes = [0u8; ::core::mem::size_of::<usize>()];
+	debug_assert_eq!(selected_route.iter().map(|p| p.get_value_msat()).sum::<u64>(), already_collected_value_msat);
+	let mut overpaid_value_msat = already_collected_value_msat - final_value_msat;
 
-	let num_permutations = payment_paths.len();
-	for _ in 0..num_permutations {
-		let mut cur_route = Vec::<PaymentPath>::new();
-		let mut aggregate_route_value_msat = 0;
+	// First, sort by the cost-per-value of the path, selecting only the paths which
+	// contribute the most per cost.
```

Comment:
Well mainly I wanted to point out the discrepancy between the comment and code. But yeah, they can of course be used somewhat analogous. That said, I assume you introduced the conversion to `u128` and the shift by `64` to compare them without the need for floats? Since `value >> cost`, could you get around that by simply sorting by `value/cost` in ascending order, rather than `cost/value` in descending order?

Reply:
Also, we're now only selecting by cost efficiency, but not by overall amount as before. Why is this better than preferring a smaller number of larger value paths, which potentially reduced the failure probability?

Reply:
> Hmm, that's a good point. For the values our scorer uses that should work alright. That said, #1600 proposes using score values of a full bitcoin, so then we'd be liable to hit issues again. To avoid any scorer-specific logic in the router just shifting up into a `u128` seemed the simplest, and while its gonna be really slow on most systems, it shouldn't matter much here, its at the end of the router and only across a handful of paths.

Ah I see. Yeah, it works as it is.

> Hmm, yea, that's a valid point - we may prefer to send a few large parts rather than a number of small parts. Still, that seems like something that should be up to the scorer - the router should seek to reduce the total cost in scorer+fee terms of the paths it selects, if the scorer wants to prefer larger parts, then great, if not, that's its problem.

Hum, I agree it's probably best to keep the routing logic as simple as possible and let the Scorer handle it going forward. And hopefully it works as intended and optimizes for failure probability.